### DISCUSSION TITLE
Idea for quickly testing on real hardware

### DISCUSSION THREAD

**Discussion by ChocolateLoverRaj** [phil-opp/blog_os#1366] [General]
## Update
This og idea is not very good. I have two better ideas:

### Network boot
Use a USB to Ethernet adapter, plug in an ethernet cable from the test device to the development device. Boot through network (I'm not sure exactly how I will do this, but it should work. It may need PXE booting to be enabled in coreboot / other firmware).

### Copy file through Linux
- Have a disk connected to the test device where you will keep the image
- Boot into Linux on the test device
- Copy the file with `rsync` or something
- Reboot into your OS

## My idea (do not use)
This is how I currently test on real hardware (based on edition 3):
- I plug in my Ventoy USB drive to the computer that I code on
- I copy the `.img` file that `build.rs` created using `bootloader` to the Ventoy
- I unplug the Ventoy USB and plug it into my Chromebook (`robo360`) which is running MrChromebox UEFI firmware.
- I boot Ventoy and select `my os.img`

I have an idea that won't need constant plugging / unplugging and file copying:
- A Rasbperry Pi Pico acts as a storage device
- On the test device, boot from the RPI Pico.
- The RPI Pico is connected to the development device using UART connected to a UART to USB converter or another RPI Pico, which is connected to the development device through USB.
- The RPI Pico asks the development device to provide the section of the disk that the UEFI firmware is asking for.
- A Rust program on the development device communicates with the RPI Pico and reads from the `my os.img` file. 
- The `my os.img` file gets streamed from the development device to the RPI Pico to the test device.

I haven't actually tried this yet since I can use QEMU for everything for now. I have a few questions:
- Would this work? Can the UEFI firmware boot from a RPI Pico which is pretending to be a flash drive?
- At what kernel size would this method be too slow? Since the RP2040 can only do 12Mbps.
- Are there any better ways of testing without so many manual steps?

**Comment by tpdenk** [comment]
IDE (which is probably the first hard drive driver that you implement) has a max speed of 5MBps (with the most basic setup), which is fast enough for probably the first few months/years of development, until you *need* something faster because files become bigger.

12MBps should be plenty IMO.

## PXE
You could look into network booting, like PXE. It's something I've been thinking about lot about, but not actually done. I'd be interested in the results if you go for it (however the setup would look like).

## SCSI
You could probably also boot from a SCSI device. I imagine the setup quite difficult, but this would be my plan b if the 12MBps really isn't enough.
